// @metaplex-foundation/js/dist/esm/plugins/candyMachineModule/CandyMachinesClient.mjs
import { CandyMachinesBuildersClient } from './CandyMachinesBuildersClient.mjs';
import { createCandyMachineOperation } from './createCandyMachine.mjs';
import { toCandyMachineConfigsFromJson } from './CandyMachineJsonConfigs.mjs';
import { findCandyMachineByAddressOperation } from './findCandyMachineByAddress.mjs';
import { findCandyMachinesByPublicKeyFieldOperation } from './findCandyMachinesByPublicKeyField.mjs';
import { findMintedNftsByCandyMachineOperation } from './findMintedNftsByCandyMachine.mjs';
import { insertItemsToCandyMachineOperation } from './insertItemsToCandyMachine.mjs';
import { updateCandyMachineOperation } from './updateCandyMachine.mjs';
import { mintCandyMachineOperation } from './mintCandyMachine.mjs';
import { CandyMachineBotTaxError } from './errors.mjs';
import { Task } from '../../utils/Task.mjs';

class CandyMachinesClient {
  constructor(metaplex) {
    this.metaplex = metaplex;
  }

  builders() {
    return new CandyMachinesBuildersClient(this.metaplex);
  }

  create(input) {
    return new Task(async scope => {
      const operation = createCandyMachineOperation(input);
      const output = await this.metaplex.operations().execute(operation, scope);
      scope.throwIfCanceled();
      const candyMachine = await this.findByAddress(output.candyMachineSigner.publicKey).run(scope);
      return { ...output,
        candyMachine
      };
    });
  }

  createFromJsonConfig(input) {
    const {
      json,
      ...otherInputs
    } = input;
    const configs = toCandyMachineConfigsFromJson(json);
    return this.create({ ...otherInputs,
      ...configs
    });
  }

  findAllByWallet(wallet, options) {
    return this.metaplex.operations().getTask(findCandyMachinesByPublicKeyFieldOperation({
      type: 'wallet',
      publicKey: wallet,
      ...options
    }));
  }

  findAllByAuthority(authority, options) {
    return this.metaplex.operations().getTask(findCandyMachinesByPublicKeyFieldOperation({
      type: 'authority',
      publicKey: authority,
      ...options
    }));
  }

  findByAddress(address, options) {
    return this.metaplex.operations().getTask(findCandyMachineByAddressOperation({
      address,
      ...options
    }));
  }

  findMintedNfts(candyMachine, options) {
    return this.metaplex.operations().getTask(findMintedNftsByCandyMachineOperation({
      candyMachine,
      ...options
    }));
  }

  insertItems(candyMachine, input) {
    return new Task(async scope => {
      const operation = insertItemsToCandyMachineOperation({
        candyMachine,
        ...input
      });
      const output = await this.metaplex.operations().execute(operation, scope);
      scope.throwIfCanceled();
      const updatedCandyMachine = await this.findByAddress(candyMachine.address).run();
      return {
        candyMachine: updatedCandyMachine,
        ...output
      };
    });
  }

  mint(candyMachine, input = {}) {
    return new Task(async scope => {
      const operation = mintCandyMachineOperation({
        candyMachine,
        ...input
      });
      const output = await this.metaplex.operations().execute(operation, scope);
      scope.throwIfCanceled();
      let nft;

      try {
        nft = await this.metaplex.nfts().findByMint(output.mintSigner.publicKey).run(scope);
      } catch (error) {
        throw new CandyMachineBotTaxError(this.metaplex.rpc().getSolanaExporerUrl(output.response.signature), error);
      }

      scope.throwIfCanceled();
      const updatedCandyMachine = await this.findByAddress(candyMachine.address).run(scope);
      return {
        nft,
        candyMachine: updatedCandyMachine,
        ...output
      };
    });
  }

  update(candyMachine, input) {
    return new Task(async scope => {
      const output = await this.metaplex.operations().execute(updateCandyMachineOperation({
        candyMachine,
        ...input
      }), scope);
      scope.throwIfCanceled();
      const updatedCandyMachine = await this.findByAddress(candyMachine.address).run();
      return {
        candyMachine: updatedCandyMachine,
        ...output
      };
    });
  }

  updateFromJsonConfig(candyMachine, input) {
    const {
      json,
      ...otherInputs
    } = input;
    const configs = toCandyMachineConfigsFromJson(json);
    return this.update(candyMachine, { ...otherInputs,
      ...configs
    });
  }

}

export { CandyMachinesClient };
