// @metaplex-foundation/js/dist/esm/plugins/candyMachineModule/updateCandyMachine.mjs
import isEqual from 'lodash.isequal';
import { createUpdateCandyMachineInstruction, createUpdateAuthorityInstruction, createSetCollectionInstruction, createRemoveCollectionInstruction } from '@metaplex-foundation/mpl-candy-machine';
import { toCandyMachineConfigs, toCandyMachineInstructionData } from './CandyMachine.mjs';
import { findCandyMachineCollectionPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { NoInstructionsToSendError } from '../../errors/SdkError.mjs';
import { isNft, isLazyNft } from '../nftModule/Nft.mjs';
import { assertSameCurrencies, SOL } from '../../types/Amount.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';
import { findMetadataPda, findMasterEditionV2Pda, findCollectionAuthorityRecordPda } from '../nftModule/pdas.mjs';
import { TokenMetadataProgram } from '../nftModule/program.mjs';

// Operation
// -----------------

const Key = 'UpdateCandyMachineOperation';
const updateCandyMachineOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const updateCandyMachineOperationHandler = {
  async handle(operation, metaplex) {
    const builder = updateCandyMachineBuilder(metaplex, operation.input);

    if (builder.isEmpty()) {
      throw new NoInstructionsToSendError(Key);
    }

    return builder.sendAndConfirm(metaplex, operation.input.confirmOptions);
  }

}; // -----------------
// Builder
// -----------------

const updateCandyMachineBuilder = (metaplex, params) => {
  const {
    candyMachine,
    authority = metaplex.identity(),
    payer = metaplex.identity(),
    newAuthority,
    newCollection: newCollectionParam,
    ...updatableFields
  } = params;
  const currentConfigs = toCandyMachineConfigs(candyMachine);
  const instructionDataWithoutChanges = toCandyMachineInstructionData(candyMachine.address, currentConfigs);
  const instructionData = toCandyMachineInstructionData(candyMachine.address, { ...currentConfigs,
    ...updatableFields
  });
  const {
    data,
    wallet,
    tokenMint
  } = instructionData;
  const shouldSendUpdateInstruction = !isEqual(instructionData, instructionDataWithoutChanges);
  const shouldSendUpdateAuthorityInstruction = !!newAuthority && !newAuthority.equals(authority.publicKey);
  const newCollection = newCollectionParam && (isNft(newCollectionParam) || isLazyNft(newCollectionParam)) ? newCollectionParam.mintAddress : newCollectionParam;
  const sameCollection = newCollection && candyMachine.collectionMintAddress && candyMachine.collectionMintAddress.equals(newCollection);
  const shouldSendSetCollectionInstruction = !!newCollection && !sameCollection;
  const shouldSendRemoveCollectionInstruction = !shouldSendSetCollectionInstruction && newCollection === null && candyMachine.collectionMintAddress !== null;
  const updateInstruction = createUpdateCandyMachineInstruction({
    candyMachine: candyMachine.address,
    authority: authority.publicKey,
    wallet
  }, {
    data
  });

  if (tokenMint) {
    updateInstruction.keys.push({
      pubkey: tokenMint,
      isWritable: false,
      isSigner: false
    });
  } else if (params.price) {
    assertSameCurrencies(params.price, SOL);
  }

  return TransactionBuilder.make() // Update data.
  .when(shouldSendUpdateInstruction, builder => {
    var _params$updateInstruc;

    return builder.add({
      instruction: updateInstruction,
      signers: [authority],
      key: (_params$updateInstruc = params.updateInstructionKey) !== null && _params$updateInstruc !== void 0 ? _params$updateInstruc : 'update'
    });
  }) // Update authority.
  .when(shouldSendUpdateAuthorityInstruction, builder => {
    var _params$updateAuthori;

    return builder.add({
      instruction: createUpdateAuthorityInstruction({
        candyMachine: candyMachine.address,
        authority: authority.publicKey,
        wallet: candyMachine.walletAddress
      }, {
        newAuthority: newAuthority
      }),
      signers: [authority],
      key: (_params$updateAuthori = params.updateAuthorityInstructionKey) !== null && _params$updateAuthori !== void 0 ? _params$updateAuthori : 'updateAuthority'
    });
  }) // Set or update collection.
  .when(shouldSendSetCollectionInstruction, builder => {
    var _params$setCollection;

    const collectionMint = newCollection;
    const metadata = findMetadataPda(collectionMint);
    const edition = findMasterEditionV2Pda(collectionMint);
    const collectionPda = findCandyMachineCollectionPda(candyMachine.address);
    const collectionAuthorityRecord = findCollectionAuthorityRecordPda(collectionMint, collectionPda);
    return builder.add({
      instruction: createSetCollectionInstruction({
        candyMachine: candyMachine.address,
        authority: authority.publicKey,
        collectionPda,
        payer: payer.publicKey,
        metadata,
        mint: collectionMint,
        edition,
        collectionAuthorityRecord,
        tokenMetadataProgram: TokenMetadataProgram.publicKey
      }),
      signers: [payer, authority],
      key: (_params$setCollection = params.setCollectionInstructionKey) !== null && _params$setCollection !== void 0 ? _params$setCollection : 'setCollection'
    });
  }) // Remove collection.
  .when(shouldSendRemoveCollectionInstruction, builder => {
    var _params$removeCollect;

    const collectionMint = candyMachine.collectionMintAddress;
    const metadata = findMetadataPda(collectionMint);
    const collectionPda = findCandyMachineCollectionPda(candyMachine.address);
    const collectionAuthorityRecord = findCollectionAuthorityRecordPda(collectionMint, collectionPda);
    return builder.add({
      instruction: createRemoveCollectionInstruction({
        candyMachine: candyMachine.address,
        authority: authority.publicKey,
        collectionPda,
        metadata,
        mint: collectionMint,
        collectionAuthorityRecord,
        tokenMetadataProgram: TokenMetadataProgram.publicKey
      }),
      signers: [authority],
      key: (_params$removeCollect = params.removeCollectionInstructionKey) !== null && _params$removeCollect !== void 0 ? _params$removeCollect : 'removeCollection'
    });
  });
};

export { updateCandyMachineBuilder, updateCandyMachineOperation, updateCandyMachineOperationHandler };
