// @metaplex-foundation/js/dist/esm/plugins/nftModule/plugin.mjs
import { cusper } from '@metaplex-foundation/mpl-token-metadata';
import { TokenMetadataProgram } from './program.mjs';
import { TokenMetadataGpaBuilder } from './gpaBuilders.mjs';
import { NftClient } from './NftClient.mjs';
import { createNftOperation, createNftOperationHandler } from './createNft.mjs';
import { findNftByMintOperation, findNftByMintOperationHandler } from './findNftByMint.mjs';
import { findNftsByCreatorOperation, findNftsByCreatorOperationHandler } from './findNftsByCreator.mjs';
import { findNftsByMintListOperation, findNftsByMintListOperationHandler } from './findNftsByMintList.mjs';
import { findNftsByOwnerOperation, findNftsByOwnerOperationHandler } from './findNftsByOwner.mjs';
import { findNftsByUpdateAuthorityOperation, findNftsByUpdateAuthorityOperationHandler } from './findNftsByUpdateAuthority.mjs';
import { findMintWithMetadataByAddressOperation, findMintWithMetadataByAddressOperationHandler } from './findMintWithMetadataByAddress.mjs';
import { findMintWithMetadataByMetadataOperation, findMintWithMetadataByMetadataOperationHandler } from './findMintWithMetadataByMetadata.mjs';
import { findTokenWithMetadataByAddressOperation, findTokenWithMetadataByAddressOperationHandler } from './findTokenWithMetadataByAddress.mjs';
import { findTokenWithMetadataByMetadataOperation, findTokenWithMetadataByMetadataOperationHandler } from './findTokenWithMetadataByMetadata.mjs';
import { findTokenWithMetadataByMintOperation, findTokenWithMetadataByMintOperationHandler } from './findTokenWithMetadataByMint.mjs';
import { loadMetadataOperation, loadMetadataOperationHandler } from './loadMetadata.mjs';
import { loadNftOperation, loadNftOperationHandler } from './loadNft.mjs';
import { printNewEditionOperation, printNewEditionOperationHandler } from './printNewEdition.mjs';
import { updateNftOperation, updateNftOperationHandler } from './updateNft.mjs';
import { uploadMetadataOperation, uploadMetadataOperationHandler } from './uploadMetadata.mjs';
import { useNftOperation, useNftOperationHandler } from './useNft.mjs';

const nftModule = () => ({
  install(metaplex) {
    // Token Metadata Program.
    metaplex.programs().register({
      name: 'TokenMetadataProgram',
      address: TokenMetadataProgram.publicKey,
      errorResolver: error => cusper.errorFromProgramLogs(error.logs, false),
      gpaResolver: metaplex => new TokenMetadataGpaBuilder(metaplex, TokenMetadataProgram.publicKey)
    }); // Operations.

    const op = metaplex.operations();
    op.register(createNftOperation, createNftOperationHandler);
    op.register(findMintWithMetadataByAddressOperation, findMintWithMetadataByAddressOperationHandler);
    op.register(findMintWithMetadataByMetadataOperation, findMintWithMetadataByMetadataOperationHandler);
    op.register(findNftByMintOperation, findNftByMintOperationHandler);
    op.register(findNftsByCreatorOperation, findNftsByCreatorOperationHandler);
    op.register(findNftsByMintListOperation, findNftsByMintListOperationHandler);
    op.register(findNftsByOwnerOperation, findNftsByOwnerOperationHandler);
    op.register(findNftsByUpdateAuthorityOperation, findNftsByUpdateAuthorityOperationHandler);
    op.register(findTokenWithMetadataByAddressOperation, findTokenWithMetadataByAddressOperationHandler);
    op.register(findTokenWithMetadataByMetadataOperation, findTokenWithMetadataByMetadataOperationHandler);
    op.register(findTokenWithMetadataByMintOperation, findTokenWithMetadataByMintOperationHandler);
    op.register(loadMetadataOperation, loadMetadataOperationHandler);
    op.register(loadNftOperation, loadNftOperationHandler);
    op.register(printNewEditionOperation, printNewEditionOperationHandler);
    op.register(updateNftOperation, updateNftOperationHandler);
    op.register(uploadMetadataOperation, uploadMetadataOperationHandler);
    op.register(useNftOperation, useNftOperationHandler);

    metaplex.nfts = function () {
      return new NftClient(this);
    };
  }

});

export { nftModule };
