// @metaplex-foundation/js/dist/esm/plugins/auctionHouseModule/Bid.mjs
import { isMintWithMetadata } from '../nftModule/Metadata.mjs';
import { amount, lamports } from '../../types/Amount.mjs';
import { Pda } from '../../types/Pda.mjs';
import { toBigNumber } from '../../types/BigNumber.mjs';
import { toDateTime, toOptionDateTime } from '../../types/DateTime.mjs';

const toBid = (account, auctionHouseModel, metadataModel) => {
  const lazyBid = toLazyBid(account, auctionHouseModel);
  return { ...lazyBid,
    model: 'bid',
    lazy: false,
    ...(isMintWithMetadata(metadataModel) ? {
      mint: metadataModel,
      tokens: amount(lazyBid.tokens, metadataModel.currency),
      isPublic: true
    } : {
      token: metadataModel,
      tokens: amount(lazyBid.tokens, metadataModel.mint.currency),
      isPublic: false
    })
  };
};
const toLazyBid = (account, auctionHouseModel) => {
  return {
    model: 'bid',
    lazy: true,
    auctionHouse: auctionHouseModel,
    tradeStateAddress: new Pda(account.data.tradeState, account.data.tradeStateBump),
    bookkeeperAddress: account.data.bookkeeper,
    buyerAddress: account.data.buyer,
    metadataAddress: account.data.metadata,
    tokenAddress: account.data.tokenAccount,
    receiptAddress: new Pda(account.data.tradeState, account.data.tradeStateBump),
    purchaseReceiptAddress: account.data.purchaseReceipt,
    isPublic: Boolean(account.data.tokenAccount),
    // Data.
    price: auctionHouseModel.isNative ? lamports(account.data.price) : amount(account.data.price, auctionHouseModel.treasuryMint.currency),
    tokens: toBigNumber(account.data.tokenSize),
    createdAt: toDateTime(account.data.createdAt),
    canceledAt: toOptionDateTime(account.data.canceledAt)
  };
};

export { toBid, toLazyBid };
