// @metaplex-foundation/js/dist/cjs/plugins/candyMachineModule/CandyMachineJsonConfigs.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplCandyMachine = require('@metaplex-foundation/mpl-candy-machine');
var Amount = require('../../types/Amount.cjs');
var PublicKey = require('../../types/PublicKey.cjs');
var BigNumber = require('../../types/BigNumber.cjs');
var DateTime = require('../../types/DateTime.cjs');

const toCandyMachineConfigsFromJson = config => {
  var _config$creators, _config$symbol, _config$maxEditionSup;

  const configCreators = (_config$creators = config.creators) !== null && _config$creators !== void 0 ? _config$creators : [{
    address: config.solTreasuryAccount,
    verified: false,
    share: 100
  }];
  return {
    wallet: config.splToken && config.splTokenAccount ? PublicKey.toPublicKey(config.splTokenAccount) : PublicKey.toPublicKey(config.solTreasuryAccount),
    tokenMint: config.splToken && config.splTokenAccount ? PublicKey.toPublicKey(config.splToken) : null,
    price: Amount.sol(config.price),
    symbol: (_config$symbol = config.symbol) !== null && _config$symbol !== void 0 ? _config$symbol : '',
    sellerFeeBasisPoints: config.sellerFeeBasisPoints,
    maxEditionSupply: BigNumber.toBigNumber((_config$maxEditionSup = config.maxEditionSupply) !== null && _config$maxEditionSup !== void 0 ? _config$maxEditionSup : 0),
    isMutable: !config.noMutable,
    retainAuthority: !config.noRetainAuthority,
    goLiveDate: DateTime.toOptionDateTime(config.goLiveDate),
    itemsAvailable: BigNumber.toBigNumber(config.number),
    endSettings: config.endSettings ? config.endSettings.endSettingType === 'date' ? {
      endSettingType: mplCandyMachine.EndSettingType.Date,
      date: DateTime.toDateTime(config.endSettings.value)
    } : {
      endSettingType: mplCandyMachine.EndSettingType.Amount,
      number: BigNumber.toBigNumber(config.endSettings.value)
    } : null,
    hiddenSettings: config.hiddenSettings ? { ...config.hiddenSettings,
      hash: Buffer.from(config.hiddenSettings.hash, 'utf8').toJSON().data
    } : null,
    whitelistMintSettings: config.whitelistMintSettings ? { ...config.whitelistMintSettings,
      mode: config.whitelistMintSettings.mode === 'burnEveryTime' ? mplCandyMachine.WhitelistMintMode.BurnEveryTime : mplCandyMachine.WhitelistMintMode.NeverBurn,
      mint: PublicKey.toPublicKey(config.whitelistMintSettings.mint),
      discountPrice: Amount.sol(config.whitelistMintSettings.discountPrice)
    } : null,
    gatekeeper: config.gatekeeper ? { ...config.gatekeeper,
      network: PublicKey.toPublicKey(config.gatekeeper.gatekeeperNetwork)
    } : null,
    creators: configCreators.map(creatorConfig => ({ ...creatorConfig,
      address: PublicKey.toPublicKey(creatorConfig.address)
    }))
  };
};

exports.toCandyMachineConfigsFromJson = toCandyMachineConfigsFromJson;
