// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/accounts.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplAuctionHouse = require('@metaplex-foundation/mpl-auction-house');
var Account = require('../../types/Account.cjs');

const parseAuctionHouseAccount = Account.getAccountParsingFunction(mplAuctionHouse.AuctionHouse);
const toAuctionHouseAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.AuctionHouse);
const parseListingReceiptAccount = Account.getAccountParsingFunction(mplAuctionHouse.ListingReceipt);
const toListingReceiptAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.ListingReceipt);
const parseBidReceiptAccount = Account.getAccountParsingFunction(mplAuctionHouse.BidReceipt);
const toBidReceiptAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.BidReceipt);

exports.parseAuctionHouseAccount = parseAuctionHouseAccount;
exports.parseBidReceiptAccount = parseBidReceiptAccount;
exports.parseListingReceiptAccount = parseListingReceiptAccount;
exports.toAuctionHouseAccount = toAuctionHouseAccount;
exports.toBidReceiptAccount = toBidReceiptAccount;
exports.toListingReceiptAccount = toListingReceiptAccount;
