// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/AuctionHouse.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assert = require('../../utils/assert.cjs');
var Pda = require('../../types/Pda.cjs');

const isAuctionHouse = value => typeof value === 'object' && value.model === 'auctionHouse';
function assertAuctionHouse(value) {
  assert["default"](isAuctionHouse(value), `Expected AuctionHouse type`);
}
const toAuctionHouse = (auctionHouseAccount, treasuryMint) => ({
  model: 'auctionHouse',
  address: new Pda.Pda(auctionHouseAccount.publicKey, auctionHouseAccount.data.bump),
  creatorAddress: auctionHouseAccount.data.creator,
  authorityAddress: auctionHouseAccount.data.authority,
  treasuryMint,
  feeAccountAddress: new Pda.Pda(auctionHouseAccount.data.auctionHouseFeeAccount, auctionHouseAccount.data.feePayerBump),
  treasuryAccountAddress: new Pda.Pda(auctionHouseAccount.data.auctionHouseTreasury, auctionHouseAccount.data.treasuryBump),
  feeWithdrawalDestinationAddress: auctionHouseAccount.data.feeWithdrawalDestination,
  treasuryWithdrawalDestinationAddress: auctionHouseAccount.data.treasuryWithdrawalDestination,
  sellerFeeBasisPoints: auctionHouseAccount.data.sellerFeeBasisPoints,
  requiresSignOff: auctionHouseAccount.data.requiresSignOff,
  canChangeSalePrice: auctionHouseAccount.data.canChangeSalePrice,
  isNative: treasuryMint.isWrappedSol
});

exports.assertAuctionHouse = assertAuctionHouse;
exports.isAuctionHouse = isAuctionHouse;
exports.toAuctionHouse = toAuctionHouse;
