// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/AuctionsClient.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var AuctionsBuildersClient = require('./AuctionsBuildersClient.cjs');
var pdas = require('./pdas.cjs');
var createAuctionHouse = require('./createAuctionHouse.cjs');
var findAuctionHouseByAddress = require('./findAuctionHouseByAddress.cjs');
var updateAuctionHouse = require('./updateAuctionHouse.cjs');
var AuctionHouseClient = require('./AuctionHouseClient.cjs');
var Task = require('../../utils/Task.cjs');

class AuctionsClient {
  constructor(metaplex) {
    this.metaplex = metaplex;
  }

  builders() {
    return new AuctionsBuildersClient.AuctionsBuildersClient(this.metaplex);
  }

  for(auctionHouse, auctioneerAuthority) {
    return new AuctionHouseClient.AuctionHouseClient(this.metaplex, auctionHouse, auctioneerAuthority);
  }

  createAuctionHouse(input) {
    return new Task.Task(async scope => {
      const output = await this.metaplex.operations().getTask(createAuctionHouse.createAuctionHouseOperation(input)).run(scope);
      scope.throwIfCanceled();
      const auctionHouse = await this.findAuctionHouseByAddress(output.auctionHouseAddress).run(scope);
      return { ...output,
        auctionHouse
      };
    });
  }

  updateAuctionHouse(auctionHouse, input) {
    return new Task.Task(async scope => {
      const output = await this.metaplex.operations().getTask(updateAuctionHouse.updateAuctionHouseOperation({
        auctionHouse,
        ...input
      })).run(scope);
      scope.throwIfCanceled();
      const updatedAuctionHouse = await this.findAuctionHouseByAddress(auctionHouse.address).run(scope);
      return { ...output,
        auctionHouse: updatedAuctionHouse
      };
    });
  }

  findAuctionHouseByAddress(address, options) {
    return this.metaplex.operations().getTask(findAuctionHouseByAddress.findAuctionHouseByAddressOperation({
      address,
      ...options
    }));
  }

  findAuctionHouseByCreatorAndMint(creator, treasuryMint, options) {
    return this.findAuctionHouseByAddress(pdas.findAuctionHousePda(creator, treasuryMint), options);
  }

}

exports.AuctionsClient = AuctionsClient;
