// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/Bid.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Metadata = require('../nftModule/Metadata.cjs');
var Amount = require('../../types/Amount.cjs');
var Pda = require('../../types/Pda.cjs');
var BigNumber = require('../../types/BigNumber.cjs');
var DateTime = require('../../types/DateTime.cjs');

const toBid = (account, auctionHouseModel, metadataModel) => {
  const lazyBid = toLazyBid(account, auctionHouseModel);
  return { ...lazyBid,
    model: 'bid',
    lazy: false,
    ...(Metadata.isMintWithMetadata(metadataModel) ? {
      mint: metadataModel,
      tokens: Amount.amount(lazyBid.tokens, metadataModel.currency),
      isPublic: true
    } : {
      token: metadataModel,
      tokens: Amount.amount(lazyBid.tokens, metadataModel.mint.currency),
      isPublic: false
    })
  };
};
const toLazyBid = (account, auctionHouseModel) => {
  return {
    model: 'bid',
    lazy: true,
    auctionHouse: auctionHouseModel,
    tradeStateAddress: new Pda.Pda(account.data.tradeState, account.data.tradeStateBump),
    bookkeeperAddress: account.data.bookkeeper,
    buyerAddress: account.data.buyer,
    metadataAddress: account.data.metadata,
    tokenAddress: account.data.tokenAccount,
    receiptAddress: new Pda.Pda(account.data.tradeState, account.data.tradeStateBump),
    purchaseReceiptAddress: account.data.purchaseReceipt,
    isPublic: Boolean(account.data.tokenAccount),
    // Data.
    price: auctionHouseModel.isNative ? Amount.lamports(account.data.price) : Amount.amount(account.data.price, auctionHouseModel.treasuryMint.currency),
    tokens: BigNumber.toBigNumber(account.data.tokenSize),
    createdAt: DateTime.toDateTime(account.data.createdAt),
    canceledAt: DateTime.toOptionDateTime(account.data.canceledAt)
  };
};

exports.toBid = toBid;
exports.toLazyBid = toLazyBid;
