// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/createBid.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var web3_js = require('@solana/web3.js');
var mplAuctionHouse = require('@metaplex-foundation/mpl-auction-house');
var pdas$1 = require('./pdas.cjs');
var Amount = require('../../types/Amount.cjs');
var pdas = require('../nftModule/pdas.cjs');
var Operation = require('../../types/Operation.cjs');
var PublicKey = require('../../types/PublicKey.cjs');
var pdas$2 = require('../tokenModule/pdas.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');
var Signer = require('../../types/Signer.cjs');

// Operation
// -----------------

const Key = 'CreateBidOperation';
const createBidOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const createBidOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const builder = await createBidBuilder(metaplex, operation.input);
    scope.throwIfCanceled();
    return builder.sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const createBidBuilder = async (metaplex, params) => {
  var _params$tokens, _params$price$basisPo, _params$price, _params$buyer, _params$authority, _params$tokenAccount, _params$bookkeeper, _params$printReceipt;

  // Data.
  const auctionHouse = params.auctionHouse;
  const tokens = (_params$tokens = params.tokens) !== null && _params$tokens !== void 0 ? _params$tokens : Amount.token(1);
  const priceBasisPoint = (_params$price$basisPo = (_params$price = params.price) === null || _params$price === void 0 ? void 0 : _params$price.basisPoints) !== null && _params$price$basisPo !== void 0 ? _params$price$basisPo : 0;
  const price = auctionHouse.isNative ? Amount.lamports(priceBasisPoint) : Amount.amount(priceBasisPoint, auctionHouse.treasuryMint.currency); // Accounts.

  const buyer = (_params$buyer = params.buyer) !== null && _params$buyer !== void 0 ? _params$buyer : metaplex.identity();
  const authority = (_params$authority = params.authority) !== null && _params$authority !== void 0 ? _params$authority : auctionHouse.authorityAddress;
  const metadata = pdas.findMetadataPda(params.mintAccount);
  const escrowPayment = pdas$1.findAuctionHouseBuyerEscrowPda(auctionHouse.address, PublicKey.toPublicKey(buyer));
  const tokenAccount = (_params$tokenAccount = params.tokenAccount) !== null && _params$tokenAccount !== void 0 ? _params$tokenAccount : params.seller ? pdas$2.findAssociatedTokenAccountPda(params.mintAccount, params.seller) : null;
  const buyerTokenAccount = pdas$2.findAssociatedTokenAccountPda(params.mintAccount, PublicKey.toPublicKey(buyer));
  const buyerTradeState = pdas$1.findAuctionHouseTradeStatePda(auctionHouse.address, PublicKey.toPublicKey(buyer), auctionHouse.treasuryMint.address, params.mintAccount, price.basisPoints, tokens.basisPoints, tokenAccount);
  const accounts = {
    wallet: PublicKey.toPublicKey(buyer),
    paymentAccount: PublicKey.toPublicKey(buyer),
    transferAuthority: PublicKey.toPublicKey(buyer),
    treasuryMint: auctionHouse.treasuryMint.address,
    metadata,
    escrowPaymentAccount: escrowPayment,
    authority: PublicKey.toPublicKey(authority),
    auctionHouse: auctionHouse.address,
    auctionHouseFeeAccount: auctionHouse.feeAccountAddress,
    buyerTradeState
  }; // Args.

  const args = {
    tradeStateBump: buyerTradeState.bump,
    escrowPaymentBump: escrowPayment.bump,
    buyerPrice: price.basisPoints,
    tokenSize: tokens.basisPoints
  }; // Sell Instruction.
  // ToDo: Add support for the auctioneerAuthority

  let buyInstruction;

  if (tokenAccount) {
    buyInstruction = mplAuctionHouse.createBuyInstruction({ ...accounts,
      tokenAccount
    }, args);
  } else {
    buyInstruction = mplAuctionHouse.createPublicBuyInstruction({ ...accounts,
      tokenAccount: buyerTokenAccount
    }, args);
  } // Signers.


  const buySigners = [buyer, authority].filter(input => !!input && Signer.isSigner(input)); // Receipt.

  const bookkeeper = (_params$bookkeeper = params.bookkeeper) !== null && _params$bookkeeper !== void 0 ? _params$bookkeeper : metaplex.identity();
  const receipt = pdas$1.findBidReceiptPda(buyerTradeState);
  const builder = TransactionBuilder.TransactionBuilder.make().setContext({
    buyerTradeState,
    tokenAccount,
    metadata,
    buyer: PublicKey.toPublicKey(buyer),
    receipt,
    bookkeeper: bookkeeper.publicKey,
    price,
    tokens
  }); // Create a TA for public bid if it doesn't exist

  if (!tokenAccount) {
    const account = await metaplex.rpc().getAccount(buyerTokenAccount);

    if (!account.exists) {
      builder.add(await metaplex.tokens().builders().createToken({
        mint: params.mintAccount,
        owner: PublicKey.toPublicKey(buyer)
      }));
    }
  }

  return builder // Create bid.
  .add({
    instruction: buyInstruction,
    signers: buySigners,
    key: 'buy'
  }) // Print the Bid Receipt.
  .when((_params$printReceipt = params.printReceipt) !== null && _params$printReceipt !== void 0 ? _params$printReceipt : true, builder => builder.add({
    instruction: mplAuctionHouse.createPrintBidReceiptInstruction({
      receipt,
      bookkeeper: bookkeeper.publicKey,
      instruction: web3_js.SYSVAR_INSTRUCTIONS_PUBKEY
    }, {
      receiptBump: receipt.bump
    }),
    signers: [bookkeeper],
    key: 'printBidReceipt'
  }));
};

exports.createBidBuilder = createBidBuilder;
exports.createBidOperation = createBidOperation;
exports.createBidOperationHandler = createBidOperationHandler;
