// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/Listing.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assert = require('../../utils/assert.cjs');
var DateTime = require('../../types/DateTime.cjs');
var Amount = require('../../types/Amount.cjs');
var Pda = require('../../types/Pda.cjs');
var BigNumber = require('../../types/BigNumber.cjs');

const isListing = value => typeof value === 'object' && value.model === 'listing' && !value.lazy;
function assertListing(value) {
  assert["default"](isListing(value), `Expected Listing type`);
}
const toListing = (account, auctionHouseModel, tokenModel) => {
  const lazyListing = toLazyListing(account, auctionHouseModel);
  return { ...lazyListing,
    model: 'listing',
    lazy: false,
    token: tokenModel,
    tokens: Amount.amount(lazyListing.tokens, tokenModel.mint.currency)
  };
};
const isLazyListing = value => typeof value === 'object' && value.model === 'listing' && value.lazy;
function assertLazyListing(value) {
  assert["default"](isLazyListing(value), `Expected LazyListing type`);
}
const toLazyListing = (account, auctionHouseModel) => {
  return {
    model: 'listing',
    lazy: true,
    auctionHouse: auctionHouseModel,
    tradeStateAddress: new Pda.Pda(account.data.tradeState, account.data.tradeStateBump),
    bookkeeperAddress: account.data.bookkeeper,
    sellerAddress: account.data.seller,
    metadataAddress: account.data.metadata,
    receiptAddress: new Pda.Pda(account.data.tradeState, account.data.tradeStateBump),
    purchaseReceiptAddress: account.data.purchaseReceipt,
    // Data.
    price: auctionHouseModel.isNative ? Amount.lamports(account.data.price) : Amount.amount(account.data.price, auctionHouseModel.treasuryMint.currency),
    tokens: BigNumber.toBigNumber(account.data.tokenSize),
    createdAt: DateTime.toDateTime(account.data.createdAt),
    canceledAt: DateTime.toOptionDateTime(account.data.canceledAt)
  };
};

exports.assertLazyListing = assertLazyListing;
exports.assertListing = assertListing;
exports.isLazyListing = isLazyListing;
exports.isListing = isListing;
exports.toLazyListing = toLazyListing;
exports.toListing = toListing;
