// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/loadBid.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Metadata = require('../nftModule/Metadata.cjs');
var assert = require('../../utils/assert.cjs');
var Operation = require('../../types/Operation.cjs');
var Amount = require('../../types/Amount.cjs');

// Operation
// -----------------

const Key = 'LoadBidOperation';
const loadBidOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const loadBidOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const {
      lazyBid,
      loadJsonMetadata = true,
      commitment
    } = operation.input;
    const bid = { ...lazyBid,
      model: 'bid',
      lazy: false
    };

    if (lazyBid.tokenAddress) {
      const tokenModel = await metaplex.nfts().findTokenWithMetadataByAddress(lazyBid.tokenAddress, {
        commitment,
        loadJsonMetadata
      }).run(scope);
      scope.throwIfCanceled();
      Metadata.assertTokenWithMetadata(tokenModel);
      assert["default"](tokenModel.metadata.address.equals(lazyBid.metadataAddress), `Token Modal metadata address must be ${lazyBid.metadataAddress}`);
      return { ...bid,
        isPublic: false,
        token: tokenModel,
        tokens: Amount.amount(lazyBid.tokens, tokenModel.mint.currency)
      };
    } else {
      const mintModel = await metaplex.nfts().findMintWithMetadataByMetadata(lazyBid.metadataAddress, {
        commitment,
        loadJsonMetadata
      }).run(scope);
      scope.throwIfCanceled();
      return { ...bid,
        isPublic: true,
        mint: mintModel,
        tokens: Amount.amount(lazyBid.tokens, mintModel.currency)
      };
    }
  }
};

exports.loadBidOperation = loadBidOperation;
exports.loadBidOperationHandler = loadBidOperationHandler;
