// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/loadListing.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Operation = require('../../types/Operation.cjs');
var Amount = require('../../types/Amount.cjs');

// -----------------
// Operation
// -----------------
const Key = 'LoadListingOperation';
const loadListingOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const loadListingOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const {
      lazyListing,
      loadJsonMetadata = true,
      commitment
    } = operation.input;
    const tokenModel = await metaplex.nfts().findTokenWithMetadataByMetadata(lazyListing.metadataAddress, lazyListing.sellerAddress, {
      commitment,
      loadJsonMetadata
    }).run(scope);
    return { ...lazyListing,
      model: 'listing',
      lazy: false,
      token: tokenModel,
      tokens: Amount.amount(lazyListing.tokens, tokenModel.mint.currency)
    };
  }
};

exports.loadListingOperation = loadListingOperation;
exports.loadListingOperationHandler = loadListingOperationHandler;
