// @metaplex-foundation/js/dist/cjs/plugins/nftModule/accounts.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var Account = require('../../types/Account.cjs');

const parseMetadataAccount = Account.getAccountParsingFunction(mplTokenMetadata.Metadata);
const toMetadataAccount = Account.getAccountParsingAndAssertingFunction(mplTokenMetadata.Metadata);
const originalOrPrintEditionAccountParser = {
  name: 'MasterEditionV1 | MasterEditionV2 | Edition',
  deserialize: (data, offset = 0) => {
    if ((data === null || data === void 0 ? void 0 : data[0]) === mplTokenMetadata.Key.MasterEditionV1) {
      return mplTokenMetadata.MasterEditionV1.deserialize(data, offset);
    } else if ((data === null || data === void 0 ? void 0 : data[0]) === mplTokenMetadata.Key.MasterEditionV2) {
      return mplTokenMetadata.MasterEditionV2.deserialize(data, offset);
    } else {
      return mplTokenMetadata.Edition.deserialize(data, offset);
    }
  }
};
const parseOriginalOrPrintEditionAccount = Account.getAccountParsingFunction(originalOrPrintEditionAccountParser);
const toOriginalOrPrintEditionAccount = Account.getAccountParsingAndAssertingFunction(originalOrPrintEditionAccountParser);
const isOriginalEditionAccount = account => {
  return 'maxSupply' in account.data;
};
const isPrintEditionAccount = account => {
  return !isOriginalEditionAccount(account);
};
const originalEditionAccountParser = {
  name: 'MasterEditionV1 | MasterEditionV2',
  deserialize: (data, offset = 0) => {
    if ((data === null || data === void 0 ? void 0 : data[0]) === mplTokenMetadata.Key.MasterEditionV1) {
      return mplTokenMetadata.MasterEditionV1.deserialize(data, offset);
    } else {
      return mplTokenMetadata.MasterEditionV2.deserialize(data, offset);
    }
  }
};
const parseOriginalEditionAccount = Account.getAccountParsingFunction(originalEditionAccountParser);
const toOriginalEditionAccount = Account.getAccountParsingAndAssertingFunction(originalEditionAccountParser);
const parsePrintEditionAccount = Account.getAccountParsingFunction(mplTokenMetadata.Edition);
const toPrintEditionAccount = Account.getAccountParsingAndAssertingFunction(mplTokenMetadata.Edition);

exports.isOriginalEditionAccount = isOriginalEditionAccount;
exports.isPrintEditionAccount = isPrintEditionAccount;
exports.parseMetadataAccount = parseMetadataAccount;
exports.parseOriginalEditionAccount = parseOriginalEditionAccount;
exports.parseOriginalOrPrintEditionAccount = parseOriginalOrPrintEditionAccount;
exports.parsePrintEditionAccount = parsePrintEditionAccount;
exports.toMetadataAccount = toMetadataAccount;
exports.toOriginalEditionAccount = toOriginalEditionAccount;
exports.toOriginalOrPrintEditionAccount = toOriginalOrPrintEditionAccount;
exports.toPrintEditionAccount = toPrintEditionAccount;
