// @metaplex-foundation/js/dist/cjs/plugins/nftModule/pdas.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var buffer = require('buffer');
var program = require('./program.cjs');
var Pda = require('../../types/Pda.cjs');
var BigNumber = require('../../types/BigNumber.cjs');

const findMetadataPda = (mint, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer()]);
};
const findMasterEditionV2Pda = (mint, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer(), buffer.Buffer.from('edition', 'utf8')]);
};
const findEditionPda = (mint, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer(), buffer.Buffer.from('edition', 'utf8')]);
};
const findEditionMarkerPda = (mint, edition, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer(), buffer.Buffer.from('edition', 'utf8'), buffer.Buffer.from(edition.div(BigNumber.toBigNumber(248)).toString())]);
};
const findCollectionAuthorityRecordPda = (mint, collectionAuthority, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer(), buffer.Buffer.from('collection_authority', 'utf8'), collectionAuthority.toBuffer()]);
};
const findUseAuthorityRecordPda = (mint, useAuthority, programId = program.TokenMetadataProgram.publicKey) => {
  return Pda.Pda.find(programId, [buffer.Buffer.from('metadata', 'utf8'), programId.toBuffer(), mint.toBuffer(), buffer.Buffer.from('user', 'utf8'), useAuthority.toBuffer()]);
};

exports.findCollectionAuthorityRecordPda = findCollectionAuthorityRecordPda;
exports.findEditionMarkerPda = findEditionMarkerPda;
exports.findEditionPda = findEditionPda;
exports.findMasterEditionV2Pda = findMasterEditionV2Pda;
exports.findMetadataPda = findMetadataPda;
exports.findUseAuthorityRecordPda = findUseAuthorityRecordPda;
